adr/ADR-15.md

```diff
@@ -0,0 +1,233 @@
+
+When the consumer already exists, for push consumers, the NATS queue subscription that was created prior to the "AddConsumer" call needs to be destroyed and replaced with the new NATS queue subscription on the consumer info's `DeliverSubject`.
+
+### Runtime
+
+#### Heartbeats
+
+If the JetStream consumer is configured with heartbeats, the server will periodically (based on the specified heartbeat interval) send hearbeats containing meta information about the stream and consumer sequences.
+
+The library should setup a timer to monitor that heartbeats from the server are properly received. No specific action is taken by the library other than notifying the user through the asynchronous error callback if heartbeats are detected missing.
+
+Also, the library should always check for JetStream heartbeat (and flow control) status messages, regardless of how the subscription was created. That is, the checks should not be conditional to the `IdleHeartbeat` option passed to a subscribe API call.
+
+The library should keep track of all messages which reply subject start with `$JS.ACK` since they contain information about the stream and consumer sequence, etc..
+
+When the library receives a message that is found to be a control message (no payload and containst the "Status" header), if the header is "100 Idle Heartbeat", then the library should check if there is any gap detected between what was the last tracked message meta information and the content of the heartbeat. If a gap is detected, an asynchronous error should be reported indicating what the gap is.
+
+Ordered consumers (should have its own ADR) will handle all that for the application, recreating the subscription as needed at the right stream sequence.
+
+**Note**: In most libraries the hearbeats messages are never presented to the user. Those are handled internally.
+
+#### JS Ack
+
+When the server sends a message to a subscriber, and if there is an AckPolicy specified other than "AckNone", then the message will have a reply subject that the library is supposed to use in order to acknowledge the message. This subject also encodes some delivery information, such as stream and consumer name, stream sequence, etc..
+
+Until now, ACK reply subject contained 9 tokens, with this layout:
+```
+$JS.ACK.<stream name>.<consumer name>.<num delivered>.<stream sequence>.<consumer sequence>.<timestamp>.<num pending>
+```
+
+This is going to change with newer servers where the number of tokens will be 10 or 11, with the total tokens being:
+```
+$JS.ACK.[<domain>.]<account hash>.<stream name>.<consumer name>.<num delivered>.<stream sequence>.<consumer sequence>.<timestamp>.<num pending>
```

Comment:
I asked this on slack but worth capturing here, when we modified these why did we chose to prepend the domain and account instead of appending?

If we appended the handling of this in clients would be significantly simpler and if we make a change now so we are not strict in the token count of the subject, any future work added to the end of these subjects would essentially remain gracefully backward compat.  Old clients will work they just won't know about the new data.

I imagine it might be because the server might want to do a single subscribe against `$JS.ACK.domain.>` or something, so wanted to find out.

Additionally how long is the account hash? Hopefully its not a 32byte hash like md5 or something, thats a LOT of bytes over time.

Reply:
Nothing against that, and Matthias is having a hard look at the moment to make sure that the decision is not undone few days later.
Still, for some libraries, adding at the end would still break anyway. For instance, the released Go client v1.11.0 is checking for the number of tokens: https://github.com/nats-io/nats.go/blob/3b1f6fcc1e1014c838036367494c3012523166b0/js.go#L1976

Reply:
Indeed, we have libraries that check the length, so some needs to change, I just want to be sure that if we change we start on an approach that will inform future changes as well

Reply:
Completely agree with that.

Reply:
@ripienaar @kozlovic the reason to push it forward is to simplify the export/import subject, as that information can make it to the front of the subscription. ie `$JS.ACK.<domain>.<account>.>`

Reply:
Correct. A new token will be added at the end and recommendations maybe to have the library not check the length will be added (if we add tokens at the end in the future).